race condition in namesys tests

Saw this in a semi-recent circleCI build
```
fatal error: concurrent map writes

goroutine 239 [running]:
runtime.throw(0x883f7f, 0x15)
	/usr/local/go/src/runtime/panic.go:566 +0x95 fp=0xc4202cdc38 sp=0xc4202cdc18
runtime.mapassign1(0x80dc80, 0xc4204d94a0, 0xc4202cdd50, 0xc4202cdd40)
	/usr/local/go/src/runtime/hashmap.go:458 +0x8ef fp=0xc4202cdd20 sp=0xc4202cdc38
gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore.(*MapDatastore).Put(0xc42002c050, 0xc4202d60a0, 0x41, 0x7d8ca0, 0xc4202c4380, 0x30, 0x0)
	/home/ubuntu/.go_workspace/src/gx/ipfs/QmRWDav6mzWseLWeYfVd5fvUKiVe9xNH29YfMF438fG364/go-datastore/basic_ds.go:28 +0x7a fp=0xc4202cdd70 sp=0xc4202cdd20
github.com/ipfs/go-ipfs/routing/offline.(*offlineRouting).PutValue(0xc420296540, 0xbaa320, 0xc4202ac060, 0xc420298a50, 0x28, 0xc4200ac180, 0xdd, 0x180, 0x0, 0x1)
	/home/ubuntu/.go_workspace/src/github.com/ipfs/go-ipfs/routing/offline/offline.go:48 +0x203 fp=0xc4202cde48 sp=0xc4202cdd70
github.com/ipfs/go-ipfs/namesys.PublishEntry(0xbaa2a0, 0xc420182d80, 0x2b22885cc878, 0xc420296540, 0xc420298a50, 0x28, 0xc420081e80, 0x0, 0x0)
	github.com/ipfs/go-ipfs/namesys/_test/_obj_test/publisher.go:248 +0x2ac fp=0xc4202cdef8 sp=0xc4202cde48
github.com/ipfs/go-ipfs/namesys.PutRecordToRouting.func1(0xc420293e00, 0xbaa2a0, 0xc420182d80, 0x2b22885cc878, 0xc420296540, 0xc420298a50, 0x28, 0xc420081e80)
	github.com/ipfs/go-ipfs/namesys/_test/_obj_test/publisher.go:188 +0x92 fp=0xc4202cdf60 sp=0xc4202cdef8
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1 fp=0xc4202cdf68 sp=0xc4202cdf60
created by github.com/ipfs/go-ipfs/namesys.PutRecordToRouting
	github.com/ipfs/go-ipfs/namesys/_test/_obj_test/publisher.go:189 +0x2de

goroutine 1 [chan receive]:
testing.(*T).Run(0xc42007e240, 0x883e44, 0x15, 0x9d78a0, 0xc42004dc01)
	/usr/local/go/src/testing/testing.go:647 +0x316
testing.RunTests.func1(0xc42007e240)
	/usr/local/go/src/testing/testing.go:793 +0x6d
testing.tRunner(0xc42007e240, 0xc42003fd98)
	/usr/local/go/src/testing/testing.go:610 +0x81
testing.RunTests(0x9d7f38, 0xc58b00, 0x8, 0x8, 0xc420443430)
	/usr/local/go/src/testing/testing.go:799 +0x2f5
testing.(*M).Run(0xc42003fef8, 0x88e73e)
	/usr/local/go/src/testing/testing.go:743 +0x85
main.main()
	github.com/ipfs/go-ipfs/namesys/_test/_testmain.go:238 +0x1c0

goroutine 17 [syscall, locked to thread]:
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:2086 +0x1

goroutine 5 [syscall]:
os/signal.signal_recv(0x0)
	/usr/local/go/src/runtime/sigqueue.go:116 +0x157
os/signal.loop()
	/usr/local/go/src/os/signal/signal_unix.go:22 +0x22
created by os/signal.init.1
	/usr/local/go/src/os/signal/signal_unix.go:28 +0x41

goroutine 6 [select]:
gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log.(*MirrorWriter).logRoutine(0xc42000ca00)
	/home/ubuntu/.go_workspace/src/gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log/writer.go:71 +0x32e
created by gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log.NewMirrorWriter
	/home/ubuntu/.go_workspace/src/gx/ipfs/QmSpJByNKFX1sCsHBEp3R73FL4NF6FnQTEGyNAXHm2GS52/go-log/writer.go:38 +0xf1

goroutine 238 [select]:
github.com/ipfs/go-ipfs/namesys.waitOnErrChan(0xbaa2a0, 0xc420182d80, 0xc420293e00, 0x0, 0x0)
	github.com/ipfs/go-ipfs/namesys/_test/_obj_test/publisher.go:209 +0x1cf
github.com/ipfs/go-ipfs/namesys.PutRecordToRouting(0xbaa2a0, 0xc420182d80, 0xbaa620, 0xc4203db830, 0xc420182c80, 0x34, 0x1, 0xed0df3e64, 0x1db9de1d, 0xc6f2c0, ...)
	github.com/ipfs/go-ipfs/namesys/_test/_obj_test/publisher.go:204 +0x470
github.com/ipfs/go-ipfs/namesys.(*ipnsPublisher).PublishWithEOL(0xc420296580, 0xbaa2e0, 0xc4200165e0, 0xbaa620, 0xc4203db830, 0xc420182c80, 0x34, 0xed0df3e64, 0x1db9de1d, 0xc6f2c0, ...)
	github.com/ipfs/go-ipfs/namesys/_test/_obj_test/publisher.go:89 +0x2a1
github.com/ipfs/go-ipfs/namesys.(*ipnsPublisher).Publish(0xc420296580, 0xbaa2e0, 0xc4200165e0, 0xbaa620, 0xc4203db830, 0xc420182c80, 0x34, 0x34, 0xc4203db8c0)
	github.com/ipfs/go-ipfs/namesys/_test/_obj_test/publisher.go:63 +0x1bf
github.com/ipfs/go-ipfs/namesys.(*mpns).Publish(0xc4203db840, 0xbaa2e0, 0xc4200165e0, 0xbaa620, 0xc4203db830, 0xc420182c80, 0x34, 0x0, 0xc70620)
	github.com/ipfs/go-ipfs/namesys/_test/_obj_test/namesys.go:112 +0xcd
github.com/ipfs/go-ipfs/namesys.TestPublishWithCache0(0xc42007e600)
	/home/ubuntu/.go_workspace/src/github.com/ipfs/go-ipfs/namesys/namesys_test.go:91 +0x25a
testing.tRunner(0xc42007e600, 0x9d78a0)
	/usr/local/go/src/testing/testing.go:610 +0x81
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:646 +0x2ec
FAIL	github.com/ipfs/go-ipfs/namesys	0.151s
make: *** [coverage/unitcover/github.com_ipfs_go-ipfs_namesys.coverprofile] Error 1
```

--- a/namesys/namesys_test.go
+++ b/namesys/namesys_test.go
@@ -12,6 +12,7 @@ import (
 
 	ci "gx/ipfs/QmP1DfoUjiWH2ZBo1PBH6FupdBucbDepx3HpWmEY6JMUpY/go-libp2p-crypto"
 	ds "gx/ipfs/QmVSase1JP7cq9QkPT46oNwdp9pT6kBkG3oqS14y3QcZjG/go-datastore"
+	dssync "gx/ipfs/QmVSase1JP7cq9QkPT46oNwdp9pT6kBkG3oqS14y3QcZjG/go-datastore/sync"
 )
 
 type mockResolver struct {
@@ -76,7 +77,7 @@ func TestNamesysResolution(t *testing.T) {
 }
 
 func TestPublishWithCache0(t *testing.T) {
-	dst := ds.NewMapDatastore()
+	dst := dssync.MutexWrap(ds.NewMapDatastore())
 	priv, _, err := ci.GenerateKeyPair(ci.RSA, 1024)
 	if err != nil {
 		t.Fatal(err)

